### PR TITLE
Update demo mobx for domvm-mobx v1.0

### PR DESCRIPTION
**Pull request by jlgrall**
Discussion: https://github.com/domvm/domvm-mobx/issues/3

### CODE REVIEW

**Comment by leeoniya** [comment]
@jlgrall 

can you squash these into a single commit?

looks good otherwise.

thanks!

**Comment by jlgrall** [comment]
What do you think about adding this in the `domvm/demos/playground/demos/mobx/mobx.html`:
```html
<-- Easily test different versions of MobX by changing the above src to:
https://cdnjs.cloudflare.com/ajax/libs/mobx/5.9.0/mobx.umd.js -->
```

**Comment by leeoniya** [comment]
then we'd have to add the same to flyd, freezer, etc., right? lol. if someone already found the html panel, hopefully they're able to figure this out. let's keep it clean.

**Comment by jlgrall** [comment]
ok

**Comment by jlgrall** [comment]
wait

**Comment by jlgrall** [comment]
Ready. (I just improved the comment about keying views in the `UsersListView`, because that is a sensitive part)

**Comment by leeoniya** [comment]
Sorry, a couple more nits.

> // We want the same checks as domvm:

not technically true, since domvm's test is `val != null`, not `val != undefined`.

if you can, please switch all cases of `undefined` assignment & testing to `== null`, which also covers `undefined`. another benefit is that `null` cannot be redefined, so you don't need to have `undefined` as the third arg to the closure. it's also shorter and minifies better.

i prefer to use 2 lines for single expression loops & conditionals [1].

```js
while (parVm = parVm.parent()) depth++;
if (_willUnmount) _willUnmount.apply(this, arguments);
if (vm._mobxObserver.stale && vm.node != null) vm.redraw(true);
```

better:

```js
while (parVm = parVm.parent())
	depth++;

if (_willUnmount)
	_willUnmount.apply(this, arguments);

// or
_willUnmount && _willUnmount.apply(this, arguments);

if (vm._mobxObserver.stale && vm.node != null)
	vm.redraw(true);
```

(and other places)

**Comment by jlgrall** [comment]
[I am away until next week.]

**Comment by jlgrall** [comment]
> Sorry, a couple more nits.

No problem, you are welcome.

> not technically true, since domvm's test is `val != null`, not `val != undefined`.

Actually it is only literally true. Technically, `val != null` is the same as `val != undefined`. So I do the same check, while avoiding the use of `null`.

My `undefined` is safe because nobody can redefine it. And it minifies to 1 character which is shorter than `null`.

There are actually multiple advantages of never using `null` (except when interacting directly with the DOM, which is rare):
- simpler tests and conditions, less risks to forget that a value could be `null` when you expect `undefined` (and vice-versa).
- JIT seeing less types of values leading to simpler and faster optimized code.
- code simplification because `undefined` is naturally present in many cases, so no need to introduce `null`: undefined variable, undefined property, undefined function argument.
- reduced use of  loose equality (`==` or `!=`): strict equality `===` is faster and less prone to bugs (especially for javascript beginners).
- having 2 types for null/undefined is a mistake in the conception of the language. We can use only one, and `undefined` is better.
- maybe others points that I forgot… It was a subject discussed on different blogs a few years ago, but I cannot find the links anymore.

I know that it doesn't make much of a speed difference to only use `undefined`, but when considering that it also makes it easier to read, debug and maintain my code, and also with the small code size improvement, I think it i a good habit to not use `null` when possible.

> i prefer to use 2 lines for single expression loops & conditionals

I am curious about this. I have always read that it should be either a conditional block or that it should be on one line, and that if you mix the both, you may introduce bugs for example when you add an instruction but you forget to add the brackets (so that the second instruction ends up outside of the condition by mistake). I have seen your way in other projects before, but I never understood if there are any reasons for it.

**Comment by lawrence-dol** [comment]
My two cents.

Loose testing for `null` and `undefined` are not the same. Null includes undefined, but undefined does not include null. Logically they are not the same either: undefined means the value was not specified (omitted), while null means a value was specified, but it is null. You may logically want to differentiate in some circumstances, but in most you want to consider null to be the same as omitted. Testing for `!=null` is nearly always the better practice. You claim that never using `null` has advantanges only at the language design level -- if the language allows null, you ***must*** account for its possibility.

> less risks to forget that a value could be `null` when you expect `undefined`

Exactly why you must test `!=null`; using `!=undefined` will miss cases where the value is `null`.

> JIT seeing less types of values leading to simpler and faster optimized code.

Irrelevant, and an compiler implementation concern.

> code simplification because `undefined` is naturally present in many cases, so no need to introduce `null`: undefined variable, undefined property, undefined function argument.

Irrelevant unless you can *enforce* the use of only one.

> reduced use of  loose equality (`==` or `!=`): strict equality `===` is faster and less prone to bugs (especially for javascript beginners).

Irrelevant to this specific concern. You can't just hide you head in the sand and pretend no one will ever pass or return null.

> having 2 types for null/undefined is a mistake in the conception of the language. We can use only one, and `undefined` is better.

That may or may not be true (there is utility in differentiating sometimes, particularly with omitted arguments). Regardless the language has both and you must account for both. Undefined is not better, it's narrower.

Lastly, and personally, I think Leo is plain wrong about two line conditionals, especially without always using braces (for the reasons you've stated). But it's his project and his call, so "suck it up, buttercup".  :-)

**Comment by leeoniya** [comment]
IMO, semantically, `null` and `undefined` should be treated as identical, and tested as being equivalent precisely *because* they are *always* conflated in real code. this may not be the case with Flow/Typescript codebases, but this project does not enforce the use of these typed supersets. treating both as equivalent results in *fewer* bugs/surprises and accepts the fact that both can and will happen. in JS i have never relied on something being strictly `undefined` but never `null` or vice versa.

> Lastly, and personally, I think Leo is plain wrong about two line conditionals, especially without always using braces (for the reasons you've stated). But it's his project and his call, so "suck it up, buttercup". :-)

with proper indentation, i've never made a mistake.

```js
for (let i = 0; i < x.length; i++)
    a.push(x[i]);
```

this is immediately obvious, at least to me:

```js
for (let i = 0; i < x.length; i++)
    a.push(x[i]);
    a.pop();
```

FWIW, i do add braces to upper layers if there are several nested `if`s:

```js
if (foo) {
    if (bar) {
       while (++x < 10)
          z++;
    }
}
```

perhaps with a different coding style it's more of a hazard? (or more devs).

there's a lot of defensive programming that is considered best practice. not using `x = x || {}` for variable assignment, not using long `? :` ternary chains. not using `x && myFn();`. not using switch/case fall-throughs. always having braces. always having `else` branches; some people love cradled `} else {`, some demand an 80 char line width. others hate semicolons. still others consider [1] to be easily readable ;)

@jlgrall if you're still not convinced about our arguments for `undefined` / `null`, then i'm not going to press it - it doesn't make the code harder to read. however, i find inline loop & conditional bodies extremely difficult to grok, so please change those to use 2 lines (or 3 if you prefer to have the braces).

[1] https://github.com/domvm/domvm/issues/203#issuecomment-424910043

**Comment by jlgrall** [comment]
Hello, and thanks for the inputs.

I appreciate working with you guys and I am always open to suggestions. I still want to maintain this project myself for the time being, so I take your messages as inputs and suggestions. If for some reason I stop maintaining it, then go ahead and rewrite it as you please to make your work easier. It won't be hard because this is just a small library.


## Some words on null/undefined

> Loose testing for null and undefined are not the same

In which case are they not the same? They are the same: `val == null` and `val == undefined` have the same result for any `val`.

>  if the language allows null, you ***must*** account for its possibility

It looks like you are misunderstanding what I wrote. It is not like I say that my API accepts either `undefined` or `null` and then in my code I only check for `undefined`. This would obviously introduce many bugs and would be bad, I agree.

What I say is: in your function signatures, your object properties and your API, only use `undefined`. Then if someone passes or assigns `null` instead of `undefined`, it is treated as a type error, not as a possible value. Just the same as if you say that a function only accepts an integer, then, receiving an empty string would be a type error (even though `"" == 0`). And you usually don't even check that the argument is actually a number, excepted when interfacing with outside code. Here it is the same: you expect only `undefined` so you can simplify your code, and most of the time you won't even check that it is not `null` because the functions and the data manipulated are private anyway so you cannot receive `null` if you don't use it anywhere else in your code.

To say it in simpler terms, I can _enforce_ the use of only `undefined` just the same way that I can enforce that an argument or a property is only an integer.

Also there are still cases where I use `null` but it is very rare and usually it can be avoided with a better design or a change in function signatures. For omitted arguments, I use `arguments.length`, which also distinguishes the case where the argument `undefined` is passed which can be tricky. For not defined properties I use `if (prop in object) …`.

## Two line conditionals

If you find inline conditionals hard to read, I could probably rewrite them on 3 lines with braces.

About the two line conditionals, I can understand that when you get used to it, it would make the code easier/faster to read. I think I can try it in this project, and when I need to come back to the codebase, to update or to fix it, I will see how it works.

## Conclusions

I am going to keep using only `undefined` (without neglecting to account for possible `null` values as I already do).

I am going to try the 2 lines conditionals without braces for now.

**Comment by jlgrall** [comment]
Ready.

File: demos/lib/domvm-mobx.js [hunk 1]
```diff
@@ -117,24 +121,28 @@ function becameStale(vm) {
...
 function eq(vm) {
-	var observerData = vm.mobxObserver;
+	var observerData = vm._mobxObserver;
 	
-	if (observerData.stale) return false;	// Re-render.
-	else if(observerData.eq) return observerData.eq.apply(this, arguments); // Let diff() choose.
-	else return true;	// By default: no re-render.
+	if (observerData.stale)
+		return false;	// Re-render.
+	else if(observerData.eq)
```
Comment (leeoniya) [inline demos/lib/domvm-mobx.js]: @jlgrall i added a follw-up commit to add the space between the `if(`

Comment (jlgrall) [reply demos/lib/domvm-mobx.js]: Thx for telling me.
